mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,43 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
-        if (agg == null) {
-          agg = Array.fill(vec.size)(
-            new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))
-        }
-        require(vec.size == agg.length,
-          s"Number of dimensions must be ${agg.length} but got ${vec.size}")
-        var i = 0
-        while (i < vec.size) {
-          agg(i) = agg(i).insert(vec(i))
-          i += 1
-        }
-      }
-
-      if (agg == null) {
-        Iterator.empty
-      } else {
-        Iterator.single(agg.map(_.compress))
-      }
-    }.treeReduce { (agg1, agg2) =>
-      require(agg1.length == agg2.length)
-      var i = 0
-      while (i < agg1.length) {
-        agg1(i) = agg1(i).merge(agg2(i))
-        i += 1
-      }
-      agg1
-    }
+    val vectors = dataset.select($(inputCol)).rdd.map { case Row(vec: Vector) => vec }
+    val numFeatures = MetadataUtils.getNumFeatures(dataset.schema($(inputCol)))
+      .getOrElse(vectors.first().size)
 
-    val (range, median) = summaries.map { s =>
-      (s.query($(upper)).get - s.query($(lower)).get,
-        s.query(0.5).get)
-    }.unzip
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    val collected = vectors.flatMap { vec =>
+      require(vec.size == numFeatures,
+        s"Number of dimensions must be $numFeatures but got ${vec.size}")
+      Iterator.range(0, numFeatures).map { i => (i, vec(i)) }
+    }.aggregateByKey(
```

Comment:
I haven't thought this through as much as you have, but, I'm concerned about whether this scales well. If I have relative low-dimensional vectors -- say just 2 dimensions -- then this will require aggregating half of all the data in just two tasks each. The original implementation doesn't depend on the dimensionality in this way. I'm not sure this is acceptable, if I'm right about it.

Unfortunately there is no way to 'map' the result of reduce or treeReduce even if it's the right tool here, because you're right that the driver doesn't need anything but medians and ranges.

How far do you get by just ensuring that treeReduce calls .compress?
And what about turning up the relative error if scale is the concern?

Reply:
I mean, I can think of hacky ideas. You can map everything to a (0, summarizer) pair, aggregateByKey, do the final aggregation in one remote executor, then _map_ out the medians and return just that. But it just pushes the problem to 1 executor. You can fake a tree reduce by mapping to (n, summarizer) pairs for 0 <= n < N << num partitions, and then do the same as above. That gets the effect of a tree reduce but lets you map the final result. I think these are fairly ugly though, so wonder if compress plus a higher error are sufficient.

Reply:
> It is a bottleneck to collect the whole Array[QuantileSummaries] from executors

I think I did not make it clear, it is not only a bottleneck for the driver, but also a bottleneck in the executors, since the `Array[QuantileSummaries]` is too large.

> If I have relative low-dimensional vectors -- say just 2 dimensions -- then this will require aggregating half of all the data in just two tasks each.

I guess the parallelism is still #partition, since according to the doc of `aggregateByKey`:
"The former operation (seqOp) is used for merging values within a partition, and the latter (combOp) is used for merging values between partitions."
So the `insert` op are executed with parallelism=#partition, and the `merge` op are with parallelism=#feature.

> You can map everything to a (0, summarizer) pair, aggregateByKey, do the final aggregation in one remote executor, then map out the medians and return just that. But it just pushes the problem to 1 executor.

Yes, this may still cause OOM when `--executor-memory` is not enough.

Reply:
Hm, then I don't think there's any way around it, as at least two summaries have to be merged at some point. Can you tune things like the relative threshold?

As I say I think this new implementation will have a more common failure mode: large data, few dimensions. Can it be adapted? like what if they key is not the dimension i, but something like `(i, partition)` and then by `i` or something?

Reply:
Good Idea! I will look into how to use partitionId.
I guess in general we can add some methods in RDD/PairRDD like `treeAggregateByKey`/`treeReduceByKey`/`aggregateByKeyWithPartitions`